### DISCUSSION TITLE
ValidateCommandBehavior should be async

### DISCUSSION THREAD

**Discussion by MostafaEsmaeili** [babaktaremi/Clean-Architecture-Template#8] [Ideas]
dear Babak
I think In the Application layer, 
`ValidateCommandBehavior` in the C# mediator should be asynchronous. 

Here is a proposal implementation: 

``` c#
if (_validators.Any())
        {
            var context = new ValidationContext<TRequest>(request);

            var validationResults = await Task.WhenAll(
                _validators.Select(v =>
                    v.ValidateAsync(context, cancellationToken)));

            var failures = validationResults
                .Where(r => r.Errors.Any())
                .SelectMany(r => r.Errors)
                .ToList();

            if (failures.Any())
                throw new ValidationException(failures);
        }
        return await next();
```

**Comment by babaktaremi** [comment]
> dear Babak
> 
> I think In the Application layer, 
> 
> `ValidateCommandBehavior` in the C# mediator should be asynchronous. 
> 
> 
> 
> Here is a proposal implementation: 
> 
> 
> 
> ``` c#
> 
> if (_validators.Any())
> 
>         {
> 
>             var context = new ValidationContext<TRequest>(request);
> 
> 
> 
>             var validationResults = await Task.WhenAll(
> 
>                 _validators.Select(v =>
> 
>                     v.ValidateAsync(context, cancellationToken)));
> 
> 
> 
>             var failures = validationResults
> 
>                 .Where(r => r.Errors.Any())
> 
>                 .SelectMany(r => r.Errors)
> 
>                 .ToList();
> 
> 
> 
>             if (failures.Any())
> 
>                 throw new ValidationException(failures);
> 
>         }
> 
>         return await next();
> 
> ```

Yep you are right... Task eliding is not a good idea
Overall I really don't like FluentValidation that much. I just put it there for people who would use it often

**Reply by MostafaEsmaeili** [reply]
> 
> Yep you are right... Task eliding is not a good idea Overall I really don't like FluentValidation that much. I just put it there for people who would use it often

so,  what is your preference for doing validations?  something like guards or so on?

**Reply by babaktaremi** [reply]
Yep
Guards are good for mostly null validations.
But I hardly come across situations where I need to use fluent validation or I have a validation logic that I use multiple times. Fluent Validation is like Automapper where you want to reduce redundant and complex code, but you need to write more complex code for their configurations.

**Comment by babaktaremi** [comment]
> dear Babak
> 
> I think In the Application layer, 
> 
> `ValidateCommandBehavior` in the C# mediator should be asynchronous. 
> 
> 
> 
> Here is a proposal implementation: 
> 
> 
> 
> ``` c#
> 
> if (_validators.Any())
> 
>         {
> 
>             var context = new ValidationContext<TRequest>(request);
> 
> 
> 
>             var validationResults = await Task.WhenAll(
> 
>                 _validators.Select(v =>
> 
>                     v.ValidateAsync(context, cancellationToken)));
> 
> 
> 
>             var failures = validationResults
> 
>                 .Where(r => r.Errors.Any())
> 
>                 .SelectMany(r => r.Errors)
> 
>                 .ToList();
> 
> 
> 
>             if (failures.Any())
> 
>                 throw new ValidationException(failures);
> 
>         }
> 
>         return await next();
> 
> ```

Yep you are right... Task eliding is not a good idea
Overall I really don't like FluentValidation that much. I just put it there for people who would use it often
